app/src/main/java/com/cdio/solitaire/controller/GameStateController.kt

```
@@ -0,0 +1,316 @@
+        if (card!!.stackID != stack.stackID) {
+            Log.e("CardNotInStack", "Card stackID differs from stack stackID")
+            return -1
+        }
+        var stackCard = stack.tail!!
+        var i = 1
+        while (stackCard.rank.ordinal != card.rank.ordinal) {
+            i++
+            stackCard = stackCard.prev!!
+        }
+        return i
+    }
+
+    fun getLastMove(): Move {
+        return gameState.moves.last()
+    }
+
+    fun updateSortedDeck(card: Card) {
+        sortedDeck[(card.suit.ordinal - 1) * 13 + card.rank.ordinal - 1] = card
+    }
+
+    /**
+     * Used to ensure consistency in tableauOrdering when performing moves.
+     */
+    private fun tableauOrdering(card: Card, targetStack: CardStack): Boolean =
+        if (targetStack.stackID !in 1..7) false
+        else when (targetStack.size) {
+            0 -> card.rank.ordinal == 13 // Only kings can go on an empty tableau
+            else -> {
+                val targetCard = targetStack.tail!!
+                val ranksMatch = targetCard.rank.ordinal - card.rank.ordinal == 1
+                val offColor = targetCard.suit.isRed() xor card.suit.isRed()
+                ranksMatch && offColor
+            }
+        }
+
+    /**
+     * Assesses whether a stack move is legal.
+     */
+    private fun verifyMoveStack(move: Move): Boolean =
+        if (move.sourceStack == null || move.targetStack == null) {
+            Log.e("IllegalMoveError", "You need to specify a sourceStack and targetStack.")
+            false
+        } else tableauOrdering(move.sourceCard!!, move.targetStack!!)
+
+    /**
+     * Assesses whether a move from foundation is legal.
+     */
+    private fun verifyMoveFromFoundation(move: Move): Boolean {
+        val foundation = move.sourceStack
+        return if (foundation == null || move.targetStack == null) {
+            Log.e("IllegalMoveError", "You need to specify a sourceStack and targetStack.")
+            false
```

Consider throwing an exception here instead of just logging the error, since a stack trace maybe could provide a lot of useful information.